### PR TITLE
PUBDEV-5557 - XGBoost dense matrix memory allocation failure

### PR DESCRIPTION
**Pull request by Pscheidl**

### CODE REVIEW

**Comment by Pscheidl** [comment]
Let'S wait for tests and @michalkurka 's review. Tests must be triple-checked ! :)

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 1]
```diff
@@ -155,14 +155,17 @@ private static int getDataRows(int[] chunks, Vec vec, int cols) {
      *
      * @param chunkIds Chunk identifier of a vector
      * @param vec      Vector containing given chunk identifiers
+     * @param weightsVector Vector with row weights, possibly null
      * @return A sum of chunk lengths. Possibly zero, if there are no chunks or the chunks are empty.
      */
-    private static int sumChunksLength(int[] chunkIds, Vec vec) {
+    private static int sumChunksLength(int[] chunkIds, Vec vec, Vec weightsVector) {
+        final long nullsInVec = weightsVector != null ? weightsVector.length() - weightsVector.nzCnt() : 0;
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: You need to use only the chunk non-zeros, not the whole Vec.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 2]
```diff
@@ -155,16 +158,71 @@ private static int getDataRows(int[] chunks, Vec vec, int cols) {
...
+          chunks = new Chunk[localChunkIDs.length];
+          for (int i = 0; i < localChunkIDs.length; i++) {
+              chunks[i] = weightsVector.chunkForChunkIdx(localChunkIDs[i]);
+          }
+    }
+
+      @Override
+      protected void map(int tid) {
+          Chunk c = chunks[tid];
+          if (!isCountedChunk(c.cidx())) return;
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: How can this happen? You are giving ids of local chunk on the input, right?

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 3]
```diff
@@ -158,12 +168,49 @@ private static int getDataRows(int[] chunks, Vec vec, int cols) {
...
+   */
+  static class ChunkZeroCounter extends MrFun<ChunkZeroCounter> {
+      int _chunkLen = 0;
+      final int[] localChunkIDs;
+      final Chunk[] chunks;
+      public ChunkZeroCounter(final Vec weightsVector, final int[] localChunkIDs) {
+          this.localChunkIDs = localChunkIDs;
+
+          chunks = new Chunk[localChunkIDs.length];
+          for (int i = 0; i < localChunkIDs.length; i++) {
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: I would move the chunk lookup to `map` - better parallelism

Comment (Pscheidl) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: Done. Moved to map.

**Comment by Pscheidl** [comment]
Besides other changes, made variables other than `_chunkLen` private and final.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java [hunk 4]
```diff
@@ -158,12 +168,48 @@ private static int getDataRows(int[] chunks, Vec vec, int cols) {
...
+
+      public LocalChunkLengthCounter(final Vec weightsVector, final int[] localChunkIDs) {
+          this.weightsVector = weightsVector;
+          this.localChunkIDs = localChunkIDs;
+    }
+
+      @Override
+      protected void map(final int tid) {
+          Chunk chunk = weightsVector.chunkForChunkIdx(tid);
+          if (chunk == null) return;
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java]: chunk should never be null

File: h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java [hunk 5]
```diff
@@ -901,6 +902,43 @@ public void testModelMetrics() {
...
   }
 
+  @Test
+  public void testCrossValidation() {
+    Scope.enter();
+    XGBoostModel denseModel = null;
+    XGBoostModel sparseModel = null;
+    try {
+      Frame tfr = Scope.track(parse_test_file("./smalldata/prostate/prostate.csv"));
+      DKV.put(tfr);
```
Comment (michalkurka) [inline h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java]: not necessary

**Comment by Pscheidl** [comment]
Reverted to previous working state with semi-linear zero counting.

**Pull request merged**